plugin/core/promise.py

```diff
@@ -1,14 +1,24 @@
-from .typing import Any, Callable, List, Tuple, Optional
+from .typing import Callable, Generic, List, Optional, Protocol, Tuple, TypeVar, Union
 import functools
 import sublime
 import threading
 
-ResolveFunc = Callable[..., None]  # Optional argument not supported in Callable so using "..."
-FullfillFunc = Callable[[ResolveFunc], None]
-ThenFunc = Callable[[Any], Any]
+T = TypeVar('T')
+T_contra = TypeVar('T_contra', contravariant=True)
+TResult = TypeVar('TResult')
 
 
-class Promise:
+class ResolveFunc(Protocol[T_contra]):
+    def __call__(self, value: Union[T_contra, 'Promise[T_contra]'] = None) -> None:
+        ...
+
+
+FullfillFunc = Callable[[T], Union[TResult, 'Promise[TResult]', None]]
```

Comment:
Does this work correctly? As in does mypy now understand `FullfillFunc[Optional[T], TResult]` ? I have a feeling that it doesn't understand the `[Optional[T], TResult]` part.

Reply:
I mean, mypy complains here:
```py
f = lambda value: value + 'sss'  # type: FullfillFunc[Optional[str], str]
```
```
plugin/core/promise.py:21: error: Unsupported left operand type for + ("None")
plugin/core/promise.py:21: note: Left operand is of type "Optional[str]"
```
Pyright doesn't complain instead.